### ISSUE TITLE
Propose for Parallel testing for Arm64

### ISSUE THREAD

**Issue by zhlhahaha** [kubevirt/kubevirt#7653]
**Is your feature request related to a problem? Please describe**:
Run e2e test parallel on Arm64 server, thus prow can assign more CI jobs to Arm64 server.

**Describe the solution you'd like**:
`quay.io/kubevirtci/bootstrap` provide a docker in docker environment. In a bootstrap image, we can use kind to create a isolated k8s cluster and run e2e tests on it.

The gaps here is that the bootstrap image does not support Arm64. And we need to maintain a separate bootstrap image for Arm64.
I had submitted a PR for build multi-arch bootstrap image https://github.com/kubevirt/project-infra/pull/1149

If the propose is accepted, I will rebase the PR.

**Additional context**:
Currently, we have two Ampere N1 bare metal servers for KubeVirt project. The server have 80 core and 250G memory. They are capable for run test parallelly.

**Status**
- [x]  crossbuild bootstrap image https://github.com/kubevirt/project-infra/pull/1149
- [x] try parallel testing in kind k8s clusters. The kind k8s clusters run in bootstrap image (A docker in docker environment)

@dhiller @rmohr I am not sure who should I cc here.

**Comment by dhiller** [comment]
@zhlhahaha there was a [comment](https://github.com/kubevirt/project-infra/pull/1149#issuecomment-942000416) that arm64 does not support nested virt. Does it support it now? 
This would be a hard requirement, otherwise we couldn't execute the same approach as we do in the other presubmits.

**Comment by zhlhahaha** [comment]
> @zhlhahaha there was a [comment](https://github.com/kubevirt/project-infra/pull/1149#issuecomment-942000416) that arm64 does not support nested virt.  Does it support it now? 

No, it not support nested virtualization yet. And that is why we try to use `kind` create k8s cluster and run e2e test parallelly.

> This would be a hard requirement, otherwise we couldn't execute the same approach as we do in the other presubmits.

Yes, maybe we can try the approach like `pull-kubevirt-e2e-kind-1.22-sriov`.

**Comment by dhiller** [comment]
Another TODO for us: In order to run the tests like we do on the prow-workloads cluster we need to integrate the cluster as a workloads cluster with prow control plane.

**Comment by EdDev** [comment]
> The gaps here is that the bootstrap image does not support Arm64. And we need to maintain a separate bootstrap image for Arm64.

I am wandering if this the only gap or there are other things behind the corner.
- kind: Have you managed to run it on your env?
- kubevirt:
  - We know that only a subset of scenarios can be run through `kind`. We unfortunately depend on some parts which are not well namespaced (e.g. storage loop devices). You should be fine as long as you do not use such "features" in the test.
  - For the existing kind runs, we disable SELinux (I guess other kernel based security services also need to be disabled, all depends on the host OS).
  - Existing jobs that run using `kind` are limited to a single instance per machine/host. We encountered several issues when trying to run two in parallel for the SR-IOV lane. There is a chance you will need to do adjustments and run stability tests when running multiple `kind` jobs on the same machine (the isolation is not as good as the one when we run it in a VM).

Lastly, I think there is a need to define an owner to the end-result content (configuration and any other integration added).
Is this something you could take on yourself? At the end of things, we will need someone to maintain all this.

**Comment by zhlhahaha** [comment]
> I am wandering if this the only gap or there are other things behind the corner.

Hi, Edward, very good question.
If we try to use `kind` to build parallel testing environment, we may have compromise on some tests (skip some tests). As nested virtualization not support on Arm64 servers, `kind` may be only way we can build parallel testing. Or another choice is just run testing on bare metal server.

> kind: Have you managed to run it on your env?

Yes, I use `kind` to build kubeVirt development environment, and run e2e tests on it.

 > We know that only a subset of scenarios can be run through kind. We unfortunately depend on some parts which are not well namespaced (e.g. storage loop devices). You should be fine as long as you do not use such "features" in the test.

It is helpful.

> For the existing kind runs, we disable SELinux (I guess other kernel based security services also need to be disabled, all depends on the host OS).

Yes, I used to met SELinux problem when running kubeVirt on ubuntu server.

> Existing jobs that run using kind are limited to a single instance per machine/host. We encountered several issues when trying to run two in parallel for the SR-IOV lane. There is a chance you will need to do adjustments and run stability tests when running multiple kind jobs on the same machine (the isolation is not as good as the one when we run it in a VM).

I have not tried to run parallel testing in multiple kind environment on Arm64. It may have some problems, but we can give it a try.

> Is this something you could take on yourself? At the end of things, we will need someone to maintain all this.

I can try my best to take care of the testing on Arm64.

**Comment by zhlhahaha** [comment]
Hi, @dhiller @EdDev 
Any future comments? If you think it is workable, I will put it on my next sprint task list.
I will do following tasks:
- [ ] crossbuild bootstrap image
- [ ] try parallel testing in kind k8s clusters. The kind k8s clusters run in bootstrap image (A docker in docker environment)

Please feel free to give any suggestion or recommendation.

**Comment by xpivarc** [comment]
> Hi, @dhiller @EdDev Any future comments? If you think it is workable, I will put it on my next sprint task list. I will do following tasks:
> 
>     * [ ]  crossbuild bootstrap image
> 
>     * [ ]  try parallel testing in kind k8s clusters. The kind k8s clusters run in bootstrap image (A docker in docker environment)
> 
> 
> Please feel free to give any suggestion or recommendation.

Hi @zhlhahaha ,
First of all, thank you for the work you did so far(It's great to see Kubevirt running on another architecture)!

I think this would be a great step forward for Kubevirt and the community. Consider this as acknowledgment from CI team and commitment to support you during the integration of the cluster to our setup. 

I think the above steps are prerequisites for joing the cluster to our prow setup. The steps afterward would be to actually join the cluster and set up some responsibilities and expectations.

@dhiller @enp0s3 @brianmcarey

**Comment by zhlhahaha** [comment]
Thanks @xpivarc

**Comment by zhlhahaha** [comment]
Update on this task.
**Verified**:
1. start 3 kind clusters on Arm64 server, and deployed kubevirt on each of them on run following command.
```
$ docker run -d -e DOCKER_IN_DOCKER_ENABLED=true -e KUBEVIRT_PROVIDER=kind-k8s-1.23 -e KUBEVIRT_NUM_NODES=3 -e KUBEVIRT_E2E_FOCUS=arm64 --privileged -v $(pwd):/workspace --name test0 -it zhlhahaha/kubevirt-bootstrap:arm64 sleep infinity
$ docker exec -it test0 /bin/bash
$ make cluster-up && make cluster-sync
```

2. parallelly running functest in the three clusters successfully and all tests have passed.
```
$ export KUBEVIRT_E2E_FOCUS=arm64
$ make build-functests && kubectl apply -f _out/manifests/testing/
$ make functests
```

3. keep the three cluster running the test over and over again for 24 hours, did not get any error.

**Issues**:
Using the krte image from `kubernetes/test-infra/images/krte` rather and the bootstrap image in the `kubevirt/project-infra/images/bootstrap`, because k8s kind cluster failed to start in the bootstrap containers with following error message:
```
 msg="RunPodSandbox for &PodSandboxMetadata{Name:kube-apiserver-kind-control-plane,Uid:b6226f318c56408f92058a8b43ee29ef,Namespace:kube-system,Attempt:0,} failed, error" error="failed to create containerd task: failed to create shim task: failed to mount rootfs component &{overlay overlay [index=off workdir=/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/79/work upperdir=/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/79/fs lowerdir=/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/7/fs]}: invalid argument: unknown"

Have not investigate this issue in deep. It seems relating to nested mount rootfs failures.
```

**Comment by zhlhahaha** [comment]
Solved:
Need to mount dockerdata/ to a host folder.

**Comment by zhlhahaha** [comment]
Record issues when try to run kubevirt CI on Arm64 cluster.

**Comment by zhlhahaha** [comment]
Issues:
- [x] 1. cross compile issue, fixed by set tag `preset-bazel-unnested: "false"`, but it seems sandbox for Arm64 missed some library.
```
10:58:50: ERROR: /home/prow/go/src/github.com/kubevirt/kubevirt/cmd/container-disk-v2alpha/BUILD.bazel:3:10: Compiling cmd/container-disk-v2alpha/main.c failed: undeclared inclusion(s) in rule '//cmd/container-disk-v2alpha:container-disk':
10:58:50: this rule is missing dependency declarations for the following files included by 'cmd/container-disk-v2alpha/main.c':
10:58:50:   '/usr/lib/gcc/aarch64-redhat-linux/8/include/stdarg.h'
10:58:50:   '/usr/lib/gcc/aarch64-redhat-linux/8/include/stddef.h'
10:58:50:   '/usr/lib/gcc/aarch64-redhat-linux/8/include/stdbool.h'
```
- [x] 2. cross compile error, fixed by https://github.com/kubevirt/kubevirt/pull/8671
```
Analyzing: target //tools/junit-merger:junit-merger (0 packages loaded, 0 targets configured)
ERROR: While resolving toolchains for target @io_bazel_rules_go//:cgo_context_data: No matching toolchains found for types @bazel_tools//tools/cpp:toolchain_type. Maybe --incompatible_use_cc_configure_from_rules_cc has been flipped and there is no default C++ toolchain added in the WORKSPACE file? See https://github.com/bazelbuild/bazel/issues/10134 for details and migration instructions.
ERROR: Analysis of target '//tools/junit-merger:junit-merger' failed; build aborted: No matching toolchains found for types @bazel_tools//tools/cpp:toolchain_type. Maybe --incompatible_use_cc_configure_from_rules_cc has been flipped and there is no default C++ toolchain added in the WORKSPACE file? See https://github.com/bazelbuild/bazel/issues/10134 for details and migration instructions.
```

- [x] 3. when start kind k8s cluster in bootstrap image, meet kubelet cgroup errors, fixed by adding SystemCgroup=true in containerd config, would write a detailed summary in the kind project, in case some other get same errors.
```
failed to run Kubelet: invalid configuration: cgroup-root ["kubelet"] doesn't exist
```

- [ ] issues when doing e2e tests in kind k8s cluster, it only happens when there is a baremetal k8s cluster. Checking now
```
13:41:28: Will run 125 of 1302 specs
13:41:38: ------------------------------
13:41:38: [SynchronizedBeforeSuite]! [PANICKED] [10.038 seconds]
13:41:38: [SynchronizedBeforeSuite] 
13:41:38: tests/tests_suite_test.go:95
13:41:38: 
13:41:38:   Test Panicked
13:41:38:   In [SynchronizedBeforeSuite] at: tests/util/util.go:20
13:41:38: 
13:41:38:   Internal error occurred: failed calling webhook "kubevirt-update-validator.kubevirt.io": failed to call webhook: Post "[https://kubevirt-operator-webhook.kubevirt.svc:443/kubevirt-validate-update?timeout=10s](https://kubevirt-operator-webhook.kubevirt.svc/kubevirt-validate-update?timeout=10s)": context deadline exceeded
13:41:38: 
13:41:38:   Full Stack Trace
13:41:38:     kubevirt.io/kubevirt/tests/util.PanicOnError(...)
13:41:38:     	tests/util/util.go:20
13:41:38:     kubevirt.io/kubevirt/tests/testsuite.AdjustKubeVirtResource()
13:41:38:     	tests/testsuite/kubevirtresource.go:114 +0xa44
13:41:38:     kubevirt.io/kubevirt/tests/testsuite.SynchronizedBeforeTestSetup()
13:41:38:     	tests/testsuite/fixture.go:95 +0xec
13:41:38: ------------------------------
13:41:38: 
13:41:38: Waiting for namespace kubevirt-test-default1 to be removed, this can take a while ...
13:41:38: Waiting for namespace kubevirt-test-alternative1 to be removed, this can take a while ...
13:41:38: Waiting for namespace kubevirt-test-operator1 to be removed, this can take a while ...
13:41:38: Waiting for namespace kubevirt-test-privileged1 to be removed, this can take a while ...
13:41:49: [SynchronizedAfterSuite]! [PANICKED] [10.414 seconds]
13:41:49: [SynchronizedAfterSuite] 
13:41:49: tests/tests_suite_test.go:97
13:41:49: 
13:41:49:   Test Panicked
13:41:49:   In [SynchronizedAfterSuite] at: tests/util/util.go:20
13:41:49: 
13:41:49:   Internal error occurred: failed calling webhook "kubevirt-update-validator.kubevirt.io": failed to call webhook: Post "[https://kubevirt-operator-webhook.kubevirt.svc:443/kubevirt-validate-update?timeout=10s](https://kubevirt-operator-webhook.kubevirt.svc/kubevirt-validate-update?timeout=10s)": context deadline exceeded
13:41:49: 
13:41:49:   Full Stack Trace
13:41:49:     kubevirt.io/kubevirt/tests/util.PanicOnError(...)
13:41:49:     	tests/util/util.go:20
13:41:49:     kubevirt.io/kubevirt/tests/testsuite.RestoreKubeVirtResource()
13:41:49:     	tests/testsuite/kubevirtresource.go:143 +0x1e8
13:41:49:     kubevirt.io/kubevirt/tests/testsuite.SynchronizedAfterTestSuiteCleanup()
13:41:49:     	tests/testsuite/fixture.go:67 +0x1c
13:41:49: ------------------------------
```

-[ ] build or tests are get freezed, jobs would not finish until timeout

**Comment by zhlhahaha** [comment]
- [ ] 13. vsock binding issue when cluster node is larger than 2 in kind environment.
virt-handler error message: 
```
{"component":"virt-handler","level":"error","msg":"Failed to bind to VSOCK port 1.","pos":"server.go:77","reason":"listen vsock host(2):1: bind: address already in use","timestamp":"2023-03-22T04:16:22.137612Z"}
```


- [ ] 14. The node label cpumanager does not updated from false to true in kind multi-node cluster, which lead to e2e test time out
```
------------------------------
[SynchronizedBeforeSuite] [FAILED] [360.108 seconds]
[SynchronizedBeforeSuite]
tests/tests_suite_test.go:96

  Timed out after 360.000s.
  Expected
      <bool>: false
  to be true
  In [SynchronizedBeforeSuite] at: tests/testsuite/kubevirtresource.go:147

  Full Stack Trace
    kubevirt.io/kubevirt/tests/testsuite.waitForSchedulableNodeWithCPUManager()
        tests/testsuite/kubevirtresource.go:147 +0xfc
    kubevirt.io/kubevirt/tests/testsuite.AdjustKubeVirtResource()
        tests/testsuite/kubevirtresource.go:136 +0xc08
    kubevirt.io/kubevirt/tests/testsuite.SynchronizedBeforeTestSetup()
        tests/testsuite/fixture.go:95 +0xec
------------------------------

Waiting for namespace kubevirt-test-default1 to be removed, this can take a while ...
Waiting for namespace kubevirt-test-alternative1 to be removed, this can take a while ...
Waiting for namespace kubevirt-test-operator1 to be removed, this can take a while ...
Waiting for namespace kubevirt-test-privileged1 to be removed, this can take a while ...

JUnit report was created: /go/src/github.com/kubevirt/kubevirt/_out/artifacts/junit.functest.xml

Ran 0 of 1416 Specs in 366.719 seconds
FAIL! -- A BeforeSuite node failed so all tests were skipped.
--- FAIL: TestTests (367.47s)
FAIL
```

**Comment by zhlhahaha** [comment]
- [ ] 17. continuously running e2e tests on a kind provider for an extended period (at least one month) can lead to network instability, resulting in some test failures. Rebooting the machine seems to resolve the issue, and the e2e tests then run smoothly.

**Comment by brianmcarey** [comment]
@zhlhahaha  as we are running the arm64 lanes in parallel are we good to close this issue?

**Comment by zhlhahaha** [comment]
> @zhlhahaha as we are running the arm64 lanes in parallel are we good to close this issue?

Hi, Brian, I record some unsolved Arm64 issues here. Should I change the title of this issue or create a new issue to record these issues?
